Implement new Wasm exception spec

danlliu opened the pull request:
#### 14eaa7f48f80bc914663fc48ad85bdcbb42688ff
<pre>
Implement new Wasm exception spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=277547">https://bugs.webkit.org/show_bug.cgi?id=277547</a>
<a href="https://rdar.apple.com/131409318">rdar://131409318</a>

Reviewed by Yusuke Suzuki.

This patch implements the new WebAssembly Exceptions proposal (with exnref).

<a href="https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md">https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md</a>

The parser now handles `try_table` and `throw_ref`, as well as `exnref`
types, in LLInt, BBQ, and OMG.

* JSTests/wasm/spec-tests/throw_ref.wast.js: Added.
(hostref):
(eq_ref):
(let.handler.get target):
(register):
(module):
(instance):
(call):
(get instance):
(exports):
(run):
(assert_malformed):
(assert_malformed_custom):
(assert_invalid):
(assert_invalid_custom):
(assert_unlinkable):
(assert_uninstantiable):
(assert_trap):
(assert_exception):
(try.f):
(catch):
(assert_exhaustion):
(assert_return):
* JSTests/wasm/spec-tests/try_table.wast.js: Added.
(hostref):
(eq_ref):
(let.handler.get target):
(register):
(module):
(instance):
(call):
(get instance):
(exports):
(run):
(assert_malformed):
(assert_malformed_custom):
(assert_invalid):
(assert_invalid_custom):
(assert_unlinkable):
(assert_uninstantiable):
(assert_trap):
(assert_exception):
(try.f):
(catch):
(assert_exhaustion):
(assert_return):
* JSTests/wasm/stress/try-table-all-ref-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-all-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-call.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-complexcatch1.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-complexcatch2.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-f32-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-f64-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-i64-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-multicatch.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-ref-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-unwind.js: Added.
(instance.exports):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/LLIntExceptions.cpp:
(JSC::LLInt::handleWasmTryTable):
* Source/JavaScriptCore/llint/LLIntExceptions.h:
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::handleWasmCatchAllThunk):
(JSC::LLInt::handleWasmTryTableThunk):
* Source/JavaScriptCore/llint/LLIntThunks.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::toValueKind):
(JSC::Wasm::BBQJITImpl::BBQJIT::addConstant):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJITImpl::BBQJIT::toB3Type):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTryTable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addThrowRef):
(JSC::Wasm::BBQJITImpl::BBQJIT::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::sizeOfType):
(JSC::Wasm::BBQJITImpl::BBQJIT::getGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::setGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStore):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitLoad):
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::sizeOfType):
(JSC::Wasm::BBQJITImpl::BBQJIT::getGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::setGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchTableImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStore):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitLoad):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::LLIntCallee::LLIntCallee):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::marshallLocation const):
(JSC::Wasm::WasmCallingConvention::numberOfStackResults const):
(JSC::Wasm::WasmCallingConvention::numberOfStackArguments const):
(JSC::Wasm::JSCallingConvention::marshallLocation const):
(JSC::Wasm::CCallingConventionArmThumb2::marshallLocation const):
(JSC::Wasm::CCallingConventionArmThumb2::numberOfStackResults const):
(JSC::Wasm::CCallingConventionArmThumb2::numberOfStackArguments const):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::addConstant):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isValidHeapTypeKind):
(JSC::Wasm::heapTypeKindAsString):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.cpp:
(JSC::Wasm::HandlerInfo::initialize):
(JSC::Wasm::HandlerInfo::handlerForIndex):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.h:
(JSC::Wasm::UnlinkedHandlerInfo::typeName const):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addTryTable):
(JSC::Wasm::IPIntGenerator::addThrowRef):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::ControlType::createTryTable):
(JSC::Wasm::LLIntGenerator::ControlType::isTryTable):
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addTryTable):
(JSC::Wasm::LLIntGenerator::addThrowRef):
(JSC::Wasm::LLIntGenerator::endTryTable):
(JSC::Wasm::LLIntGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::isTryTable):
(JSC::Wasm::OMGIRGenerator::ControlData::dump const):
(JSC::Wasm::OMGIRGenerator::ControlData::endTryTable):
(JSC::Wasm::OMGIRGenerator::ControlData::tryStart const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryEnd const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryDepth const):
(JSC::Wasm::OMGIRGenerator::ControlData::exception const):
(JSC::Wasm::OMGIRGenerator::addTryTable):
(JSC::Wasm::OMGIRGenerator::emitCatchTableImpl):
(JSC::Wasm::OMGIRGenerator::addThrowRef):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::typeKindSizeInBytes):
* Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::set):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/283954@main">https://commits.webkit.org/283954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/943048cc8a56264dd7a1649a86a2f56a7d480e6f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67832 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20471 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71888 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/18973 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18779 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54266 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12680 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70899 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58654 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34733 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39955 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16064 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17331 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60949 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61920 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73585 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67079 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11795 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15691 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61721 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/root-captured-as-different-tag.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58728 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9615 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3234 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88860 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10331 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43021 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15677 "Found 28 new JSC stress test failures: wasm.yaml/wasm/fuzz/export-function.js.wasm-bbq, wasm.yaml/wasm/spec-tests/throw_ref.wast.js.default-wasm, wasm.yaml/wasm/spec-tests/throw_ref.wast.js.wasm-bbq, wasm.yaml/wasm/spec-tests/throw_ref.wast.js.wasm-collect-continuously, wasm.yaml/wasm/spec-tests/throw_ref.wast.js.wasm-eager-jettison, wasm.yaml/wasm/spec-tests/throw_ref.wast.js.wasm-no-cjit, wasm.yaml/wasm/spec-tests/try_table.wast.js.default-wasm, wasm.yaml/wasm/spec-tests/try_table.wast.js.wasm-bbq, wasm.yaml/wasm/spec-tests/try_table.wast.js.wasm-collect-continuously, wasm.yaml/wasm/spec-tests/try_table.wast.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44098 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -491,6 +491,38 @@ MacroAssemblerCodeRef<ExceptionHandlerPtrTag> handleWasmCatchAllThunk(OpcodeSize
+        static LazyNeverDestroyed<MacroAssemblerCodeRef<ExceptionHandlerPtrTag>> codeRef;
+        static std::once_flag onceKey;
+        std::call_once(onceKey, [&] {
+            codeRef.construct(generateThunkWithJumpTo<ExceptionHandlerPtrTag>(LLInt::getCodeFunctionPtr<OperationPtrTag>(opcode), "wasm_catch_all"));
```
try_table?

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -491,6 +491,38 @@ MacroAssemblerCodeRef<ExceptionHandlerPtrTag> handleWasmCatchAllThunk(OpcodeSize
+        static LazyNeverDestroyed<MacroAssemblerCodeRef<ExceptionHandlerPtrTag>> codeRef;
+        static std::once_flag onceKey;
+        std::call_once(onceKey, [&] {
+            codeRef.construct(generateThunkWithJumpTo<ExceptionHandlerPtrTag>(LLInt::getWide16CodeFunctionPtr<OperationPtrTag>(opcode), "wasm_catch_all16"));
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -491,6 +491,38 @@ MacroAssemblerCodeRef<ExceptionHandlerPtrTag> handleWasmCatchAllThunk(OpcodeSize
+        static LazyNeverDestroyed<MacroAssemblerCodeRef<ExceptionHandlerPtrTag>> codeRef;
+        static std::once_flag onceKey;
+        std::call_once(onceKey, [&] {
+            codeRef.construct(generateThunkWithJumpTo<ExceptionHandlerPtrTag>(LLInt::getWide32CodeFunctionPtr<OperationPtrTag>(opcode), "wasm_catch_all32"));
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -2741,6 +2750,79 @@ commonWasmOp(wasm_catch_all, WasmCatchAll, macro() end, macro(ctx)
+
+.done:
+    dispatch(ctx)
+end)
```
Let's put empty line here.

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -2741,6 +2750,79 @@ commonWasmOp(wasm_catch_all, WasmCatchAll, macro() end, macro(ctx)
+
+.done:
+    dispatch(ctx)
+end)
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -2741,6 +2750,79 @@ commonWasmOp(wasm_catch_all, WasmCatchAll, macro() end, macro(ctx)
+    # don't do anything, we don't push at all just jump
+    commonCatchImpl(ctx)
+    dispatch(ctx)
+end)
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3574,6 +3596,22 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addThrow(unsigned exceptionIndex, Argum
 }
 
+PartialResult WARN_UNUSED_RETURN BBQJIT::addThrowRef(Value exception, Stack&)
+{
```
Let's put LOG_INSTRUCTION.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3789,6 +3827,23 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addEndToUnreachable(ControlEntry& entry
+            emitCatchTableImpl(entryData, target);
+
+        // we're done!
+        --m_tryCatchDepth;
```
I wonder if we should just generate these sequence at the end of the entire code so that we can make main code path tight (since exception handling happens rarely). We should have FIXME comment about that.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3239,6 +3250,107 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+        uint32_t numberOfCatches;
+        Vector<CatchHandler> targets;
+        WASM_PARSER_FAIL_IF(!parseVarUInt32(numberOfCatches), "can't get the number of catch statements for try_table"_s);
+        WASM_PARSER_FAIL_IF(numberOfCatches == std::numeric_limits<uint32_t>::max(), "try_table's number of catch targets is too big "_s, numberOfCatches);
```
I wonder if we should have smaller limits. Can you check the spec / the other implementation to see what limit is used here? And let's define that limit in WasmLimits.h.

danlliu replied:
The spec doesn't provide any limit besides `varuint32`, and it doesn't appear other engines specify any smaller limits.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3239,6 +3250,107 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+        ControlType control;
+        Stack newStack;
+
+        if (targets.size()) {
```
This check is not necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3283,6 +3395,17 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+    case ThrowRef: {
+        TypedExpression exn;
+        WASM_TRY_POP_EXPRESSION_STACK_INTO(exn, "exception reference"_s);
+        auto exnref = Type { TypeKind::RefNull, static_cast<TypeIndex>(TypeKind::Exn) };
```
Let's define `exnrefType()` in WasmFormat.h.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3599,6 +3722,11 @@ auto FunctionParser<Context>::parseUnreachableExpression() -> PartialResult
     }
 
+    case TryTable: {
+        ASSERT(false);
```
Is it correct? I think we can hit if it is defined in unreachable block.

danlliu replied:
Agh, I forgot to put in the code here, fixing that right now.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3690,6 +3818,9 @@ auto FunctionParser<Context>::parseUnreachableExpression() -> PartialResult
 
         return { };
     }
```
Add empty line.

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1303,6 +1347,11 @@ auto LLIntGenerator::addRethrow(unsigned, ControlType& data) -> PartialResult
     return { };
 }
 
+auto WARN_UNUSED_RETURN LLIntGenerator::addThrowRef(ExpressionType exception, Stack&) -> PartialResult {
```
WARN_UNUSED_RETURN is not necessary.
And need to put `{` in the next line.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4763,6 +4873,24 @@ auto OMGIRGenerator::addThrow(unsigned exceptionIndex, ArgumentList& args, Stack
     return { };
 }
 
+auto WARN_UNUSED_RETURN OMGIRGenerator::addThrowRef(ExpressionType exn, Stack&) -> PartialResult {
```
`{` needs to be next line.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -1,4 +1,4 @@
-/*
+/*d
```
Undo this one.

Constellation commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -301,6 +301,7 @@ constexpr size_t typeKindSizeInBytes(TypeKind kind)
     case TypeKind::Arrayref:
     case TypeKind::Structref:
     case TypeKind::Funcref:
+    case TypeKind::Exn:
```
Isn't it `Exnref`?

danlliu replied:
Since we have GC now, `exnref` == `(ref null exn)`, so I'm just providing the base heap type `exn` here.

Constellation commented on Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:
```diff
@@ -284,6 +284,7 @@ def atomicMemoryLog2AlignmentGenerator(filter):
         switch(kind) {
         case TypeKind::I64:
         case TypeKind::Funcref:
+        case TypeKind::Exn:
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -156,6 +156,7 @@ void JSWebAssemblyStruct::set(uint32_t fieldIndex, uint64_t argument)
     case TypeKind::Sub:
     case TypeKind::Subfinal:
     case TypeKind::Rec:
+    case TypeKind::Exn:
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -163,6 +163,7 @@ Expected<MacroAssemblerCodeRef<WasmEntryPtrTag>, BindingFailure> wasmToJS(TypeIn
                 RELEASE_ASSERT_NOT_REACHED(); // Handled above.
             case TypeKind::RefNull:
             case TypeKind::Ref:
+            case TypeKind::Exn:
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
```diff
@@ -257,6 +258,7 @@ Expected<MacroAssemblerCodeRef<WasmEntryPtrTag>, BindingFailure> wasmToJS(TypeIn
                 RELEASE_ASSERT_NOT_REACHED(); // Handled above.
             case TypeKind::RefNull:
             case TypeKind::Ref:
+            case TypeKind::Exn:
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/wasm.json:
```diff
@@ -22,6 +22,7 @@
         "i31ref":        { "type": "varint7", "value":  -20, "b3type": "B3::Void",   "width": 0 },
         "structref":     { "type": "varint7", "value":  -21, "b3type": "B3::Int64",  "width": 0 },
         "arrayref":      { "type": "varint7", "value":  -22, "b3type": "B3::Int64",  "width": 0 },
+        "exn":           { "type": "varint7", "value":  -23, "b3type": "B3::Int64",  "width": 0 },
```
ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3571,6 +3593,25 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addThrow(unsigned exceptionIndex, Argum
+        flushRegisters();
+    }
+    m_jit.move(GPRInfo::wasmContextInstancePointer, GPRInfo::argumentGPR0);
+    emitRethrowImpl(m_jit);
```
Why is it `rethrow`? If it needs to be `rethrow`, I wonder if we should rename `emitRethrowImpl` and operations to `emitThrowRefImpl` etc. and implement rethrow with them.

danlliu replied:
These two end up sharing the same path once we find the exception pointer to throw, I think it'd make sense to rename it though since this is the "main" exception spec now.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3437,6 +3441,24 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addTry(BlockSignature signature, Stack&
+    result.setTryInfo(m_callSiteIndex, m_callSiteIndex, m_tryCatchDepth);
+    currentControlData().flushAndSingleExit(*this, result, enclosingStack, true, false);
+
+    m_pendingTryTables.append(targets);
```
Instead of using `m_pendingTryeTables` member variables to track tables, how about storing it to `ControlData`?

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3793,6 +3834,25 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addEndToUnreachable(ControlEntry& entry
+        // a real landing pad
+        // FIXME: should we generate this all at the end of the code? this might help icache performance since
+        // exceptions are rare
+        auto targets = m_pendingTryTables.takeLast();
```
And instead of using `m_pendingTryTables`, getting this from `entryData`.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -3019,6 +3023,109 @@ void BBQJIT::emitCatchImpl(ControlData& dataCatch, const TypeDefinition& excepti
+        auto signature = target.exceptionSignature->template as<FunctionSignature>();
+        if (signature->argumentCount()) {
+            ScratchScope<1, 0> scratches(*this);
+            GPRReg bufferGPR = scratches.gpr(0);
```
What is the reason of not using wasmScratchGPR?

Constellation commented on Source/JavaScriptCore/wasm/WasmFormat.h:
```diff
@@ -397,6 +403,8 @@ inline const char* heapTypeKindAsString(TypeKind kind)
     case TypeKind::Nullexternref:
         return "noextern";
+    case TypeKind::Exn:
+        return "Exn";
```
This should be `exn`

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3386,6 +3397,105 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+        ControlType control;
+        Stack newStack;
+
+        for (unsigned i = 0; i < targets.size(); ++i) {
```
Use `for (auto& target : targets)`

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3430,6 +3540,17 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+    case ThrowRef: {
+        TypedExpression exn;
+        WASM_TRY_POP_EXPRESSION_STACK_INTO(exn, "exception reference"_s);
+        auto exnref = exnrefType();
```
Remove this.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3430,6 +3540,17 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+        TypedExpression exn;
+        WASM_TRY_POP_EXPRESSION_STACK_INTO(exn, "exception reference"_s);
+        auto exnref = exnrefType();
+        WASM_VALIDATOR_FAIL_IF(exn.type() != exnref, "throw_ref expected an exception reference"_s);
```
Just do,
```
WASM_VALIDATOR_FAIL_IF(exn.type() != exnrefType(), "..."_s);
```

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1175,6 +1204,21 @@ auto LLIntGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+    Ref<Label> tryLabel = newEmittedLabel();
+    Ref<Label> continuation = newLabel();
+    result = ControlType::createTryTable(signature, m_stackSize, WTFMove(tryLabel), WTFMove(continuation), m_tryDepth);
+    m_pendingTryTables.add(&std::get<ControlTryTable>(result).m_try.get(), targets);
```
Let's put the table into corresponding `ControlType` instead of using `m_pendingTryTables`.

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1434,6 +1484,80 @@ auto LLIntGenerator::endBlock(ControlEntry& entry, Stack& expressionStack) -> Pa
+    auto tryTable = std::get<ControlTryTable>(data);
+    ASSERT(m_pendingTryTables.contains(&tryTable.m_try.get()));
+    auto targets = m_pendingTryTables.get(&tryTable.m_try.get());
+    m_pendingTryTables.remove(&tryTable.m_try.get());
```
Let's get table from `data`.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4597,6 +4612,18 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
     return { };
 }
 
+auto OMGIRGenerator::addTryTable(BlockSignature signature, Stack& enclosingStack, const Vector<CatchHandler>& targets, ControlType& result, Stack& newStack) -> PartialResult {
```
`{` should be in the next line.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4597,6 +4612,18 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+    BasicBlock* continuation = m_proc.addBlock();
+    splitStack(signature, enclosingStack, newStack);
+    result = ControlData(m_proc, origin(), signature, BlockType::TryTable, m_stackSize, continuation, advanceCallSiteIndex(), m_tryCatchDepth);
+    m_pendingTryTables.append(targets);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4526,6 +4545,24 @@ auto OMGIRGenerator::addRethrow(unsigned, ControlType& data) -> PartialResult
     return { };
 }
 
+auto WARN_UNUSED_RETURN OMGIRGenerator::addThrowRef(ExpressionType exn, Stack&) -> PartialResult {
```
Ditto. `{` should be in the next line in WebKit coding style.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4330,6 +4337,18 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
     return { };
 }
 
+auto OMGIRGenerator::addTryTable(BlockSignature signature, Stack& enclosingStack, const Vector<CatchHandler>& targets, ControlType& result, Stack& newStack) -> PartialResult {
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3437,6 +3446,34 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addTry(BlockSignature signature, Stack&
+
+    result = ControlData(*this, BlockType::TryTable, signature, currentControlData().enclosedHeight() + currentControlData().implicitSlots() + enclosingStack.size() - signature->argumentCount());
+    result.setTryInfo(m_callSiteIndex, m_callSiteIndex, m_tryCatchDepth);
+    result.setTryTableTargets(targetList);
```
`WTFMove(targetList)`

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -631,6 +632,11 @@ void ControlData::setTryInfo(unsigned tryStart, unsigned tryEnd, unsigned tryCat
     m_tryCatchDepth = tryCatchDepth;
 }
 
+void ControlData::setTryTableTargets(const TargetList &targets)
```
Get `TargetList&& targets` and use
```
m_tryTableTargets = WTFMove(targets);
```

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -244,6 +248,25 @@ class OMGIRGenerator {
+        void setTryTableTargets(const TargetList& targets)
+        {
+            m_tryTableTargets = targets;
+        }
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4330,6 +4357,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+    BasicBlock* continuation = m_proc.addBlock();
+    splitStack(signature, enclosingStack, newStack);
+    result = ControlData(m_proc, origin(), signature, BlockType::TryTable, m_stackSize, continuation, advanceCallSiteIndex(), m_tryCatchDepth);
+    result.setTryTableTargets(targetList);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -289,6 +292,25 @@ class OMGIRGenerator {
+        {
+            m_tryTableTargets = targets;
+        }
+
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4597,6 +4625,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+    BasicBlock* continuation = m_proc.addBlock();
+    splitStack(signature, enclosingStack, newStack);
+    result = ControlData(m_proc, origin(), signature, BlockType::TryTable, m_stackSize, continuation, advanceCallSiteIndex(), m_tryCatchDepth);
+    result.setTryTableTargets(targetList);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4597,6 +4625,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+            targets[i].exceptionSignature,
+            targets[i].target
+        };
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4330,6 +4357,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+            targets[i].exceptionSignature,
+            targets[i].target
+        };
+    }
```
Let's do,
```
auto targetList = targets.map(
    [&](const auto& target) {
        return {
            target.type,
            target.tag,
            target.exceptionSignature,
            target.target
        };
    });
```

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1175,6 +1216,31 @@ auto LLIntGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+            targets[i].target->targetLabelForBranch().get(),
+            targets[i].target->stackSize()
+        };
+    }
```
Ditto.

Constellation replied:
Yeah, let's use `map`.

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1175,6 +1216,31 @@ auto LLIntGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+
+    Ref<Label> tryLabel = newEmittedLabel();
+    Ref<Label> continuation = newLabel();
+    result = ControlType::createTryTable(signature, m_stackSize, WTFMove(tryLabel), WTFMove(continuation), m_tryDepth, targetList);
```
Use `WTFMove(targetList)`

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -118,11 +136,17 @@ class LLIntGenerator : public BytecodeGeneratorBase<GeneratorTraits> {
             return ControlType(signature, stackSize - signature->argumentCount(), WTFMove(continuation), ControlTry { WTFMove(tryLabel), tryDepth });
         }
 
+        static ControlType createTryTable(BlockSignature signature, unsigned stackSize, Ref<Label>&& tryLabel, RefPtr<Label>&& continuation, unsigned tryDepth, Vector<ControlTryTable::TryTableTarget, 8> targets)
```
Take `Vector<ControlTryTable::TryTableTarget, 8>` as `Vector<ControlTryTable::TryTableTarget, 8>&&` and use `WTFMove(targets)`

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -578,6 +614,7 @@ class LLIntGenerator : public BytecodeGeneratorBase<GeneratorTraits> {
     const unsigned m_functionIndex { UINT_MAX };
     Vector<VirtualRegister> m_normalizedArguments;
     HashMap<Label*, Vector<SwitchEntry>> m_switches;
+    HashMap<Label*, Vector<CatchEntry>> m_tryTableTargets;
```
This is no longer necessary, right?

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1434,6 +1506,78 @@ auto LLIntGenerator::endBlock(ControlEntry& entry, Stack& expressionStack) -> Pa
+        m_stackSize = target.targetStackSize;
+
+        ResultList results;
+        if (target.type == CatchKind::Catch || target.type == CatchKind::CatchRef) {
```
Do
```
results.results.reserveInitialCapacity(target.exceptionSignature->as<FunctionSignature>()->argumentCount());
```

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4695,6 +4847,14 @@ auto OMGIRGenerator::addEndToUnreachable(ControlEntry& entry, const Stack& expre
+        // emit each handler as a new basic block
+        data.endTryTable(advanceCallSiteIndex());
+        auto targets = data.m_tryTableTargets;
+        for (auto& target : targets)
```
Do not copy here. Use,
```
for (auto& target : data.m_tryTableTargets)
```

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3793,6 +3849,25 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addEndToUnreachable(ControlEntry& entry
+        ++m_callSiteIndex;
+
+        for (auto target : targets)
+            emitCatchTableImpl(entryData, target);
```
Do not copy entryData.m_tryTableTargets.
Do,
```
++m_callSiteIndex;
for (auto& target : entryData.m_tryTableTargets)
```

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -2966,8 +2969,7 @@ void BBQJIT::emitCatchImpl(ControlData& dataCatch, const TypeDefinition& excepti
     if (exceptionSignature.as<FunctionSignature>()->argumentCount()) {
-        ScratchScope<1, 0> scratches(*this);
-        GPRReg bufferGPR = scratches.gpr(0);
+        GPRReg bufferGPR = wasmScratchGPR;
```
I like using `wasmScratchGPR` directly instead of naming it to bufferGPR since it makes using wasmScratchGPR explicit.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -3019,6 +3022,108 @@ void BBQJIT::emitCatchImpl(ControlData& dataCatch, const TypeDefinition& excepti
+    if (target.type == CatchKind::Catch || target.type == CatchKind::CatchRef) {
+        auto signature = target.exceptionSignature->template as<FunctionSignature>();
+        if (signature->argumentCount()) {
+            GPRReg bufferGPR = wasmScratchGPR;
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -3386,6 +3397,105 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
+        for (auto& catchTarget : targets) {
+            ControlType* target = catchTarget.target;
+
+            Stack results;
```
Do,
```
results.reserveInitialCapacity(target->branchTargetArity());
```

Constellation commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1434,6 +1506,78 @@ auto LLIntGenerator::endBlock(ControlEntry& entry, Stack& expressionStack) -> Pa
+    // jump past all handlers
+    WasmJmp::emit(this, end->bind(this));
+    auto oldStackSize = m_stackSize;
+    for (auto &target : targets) {
```
`auto& target`, WebKit coding style requires putting `&` to the type side.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3437,6 +3446,34 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addTry(BlockSignature signature, Stack&
+            targets[i].target
+        };
+    }
+
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/LLIntThunks.cpp:
```diff
@@ -491,6 +491,38 @@ MacroAssemblerCodeRef<ExceptionHandlerPtrTag> handleWasmCatchAllThunk(OpcodeSize
+            codeRef.construct(generateThunkWithJumpTo<ExceptionHandlerPtrTag>(LLInt::getWide32CodeFunctionPtr<OperationPtrTag>(opcode), "wasm_try_table32"));
+        });
+        return codeRef;
+    }
```
Each `call_once` is called only once, so changing `opcode` does not have effect. (if we call this function with different `opcode` then the first one gets returned)

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3437,6 +3446,34 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addTry(BlockSignature signature, Stack&
+            target.exceptionSignature,
+            target.target
+        };
+    });
```
Do one level-indent here.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3793,6 +3849,24 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addEndToUnreachable(ControlEntry& entry
+        // exceptions are rare
+        ++m_callSiteIndex;
+
+        for (auto target : entryData.m_tryTableTargets)
```
Take `auto&`

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -1087,6 +1088,28 @@ WASM_SLOW_PATH_DECL(rethrow)
+    JSWebAssemblyException* exception = reinterpret_cast<JSWebAssemblyException*>(READ(instruction.m_exception).pointer());
+
+    if (!exception)
+        WASM_THROW(Wasm::ExceptionType::NullExnReference);
```
It it actually null pointer? Or `jsNull()`? (If it is `jsNull()`, we should `READ(instruction.m_exception)` as EncodedJSValue.
Also, I saw this exception only in LLInt's throw_ref. Don't we need to do this kind of handling in BBQ and OMG?

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4330,6 +4357,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+            target.exceptionSignature,
+            target.target
+        };
+    });
```
Let's do one level indent more.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4597,6 +4625,29 @@ auto OMGIRGenerator::addTry(BlockSignature signature, Stack& enclosingStack, Con
+            target.exceptionSignature,
+            target.target
+        };
+    });
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -1087,6 +1088,29 @@ WASM_SLOW_PATH_DECL(rethrow)
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    auto instruction = pc->as<WasmThrowRef>();
+    auto exceptionPtr = READ(instruction.m_exception).pointer();
+    EncodedJSValue exception = EncodedJSValue(exceptionPtr);
```
Let's get `JSValue exception = READ(instruction.m_exception).jsValue()`

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -1087,6 +1088,29 @@ WASM_SLOW_PATH_DECL(rethrow)
+    auto exceptionPtr = READ(instruction.m_exception).pointer();
+    EncodedJSValue exception = EncodedJSValue(exceptionPtr);
+
+    if (JSValue::decode(exception) == jsNull())
```
And do `exception.isNull()`.

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -1087,6 +1088,29 @@ WASM_SLOW_PATH_DECL(rethrow)
+    if (JSValue::decode(exception) == jsNull())
+        WASM_THROW(Wasm::ExceptionType::NullExnReference);
+
+    throwException(globalObject, throwScope, reinterpret_cast<JSWebAssemblyException*>(exceptionPtr));
```
And do, throwException(globalObject, throwScope, exception)`.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3571,6 +3609,35 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addThrow(unsigned exceptionIndex, Argum
+    }
+
+    // Check for a null exception
+    m_jit.move(CCallHelpers::TrustedImmPtr(JSValue::encode(jsNull())), wasmScratchGPR);
+    auto nullexn = m_jit.branchPtr(CCallHelpers::Equal, GPRInfo::argumentGPR1, wasmScratchGPR);
```
You can just check

```
auto isNullExn = m_jit.branch64(CCallHelpers::Equal, GPRInfo::argumentGPR1, CCallHelpers::TrustedImm64(JSValue::encode(jsNull()));
```

danlliu committed:
Implement new Wasm exception spec
https://bugs.webkit.org/show_bug.cgi?id=277547
rdar://131409318

Reviewed by Yusuke Suzuki.

This patch implements the new WebAssembly Exceptions proposal (with exnref).

https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/Exceptions.md

The parser now handles `try_table` and `throw_ref`, as well as `exnref`
types, in LLInt, BBQ, and OMG.

* JSTests/wasm/spec-tests/throw_ref.wast.js: Added.
(hostref):
(eq_ref):
(let.handler.get target):
(register):
(module):
(instance):
(call):
(get instance):
(exports):
(run):
(assert_malformed):
(assert_malformed_custom):
(assert_invalid):
(assert_invalid_custom):
(assert_unlinkable):
(assert_uninstantiable):
(assert_trap):
(assert_exception):
(try.f):
(catch):
(assert_exhaustion):
(assert_return):
* JSTests/wasm/spec-tests/try_table.wast.js: Added.
(hostref):
(eq_ref):
(let.handler.get target):
(register):
(module):
(instance):
(call):
(get instance):
(exports):
(run):
(assert_malformed):
(assert_malformed_custom):
(assert_invalid):
(assert_invalid_custom):
(assert_unlinkable):
(assert_uninstantiable):
(assert_trap):
(assert_exception):
(try.f):
(catch):
(assert_exhaustion):
(assert_return):
* JSTests/wasm/stress/try-table-all-ref-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-all-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-call.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-complexcatch1.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-complexcatch2.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-f32-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-f64-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-i64-argument.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-multicatch.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-ref-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-simple.js: Added.
(instance.exports):
* JSTests/wasm/stress/try-table-unwind.js: Added.
(instance.exports):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/LLIntExceptions.cpp:
(JSC::LLInt::handleWasmTryTable):
* Source/JavaScriptCore/llint/LLIntExceptions.h:
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::handleWasmCatchAllThunk):
(JSC::LLInt::handleWasmTryTableThunk):
* Source/JavaScriptCore/llint/LLIntThunks.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::toValueKind):
(JSC::Wasm::BBQJITImpl::BBQJIT::addConstant):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJITImpl::BBQJIT::toB3Type):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTryTable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addThrowRef):
(JSC::Wasm::BBQJITImpl::BBQJIT::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::sizeOfType):
(JSC::Wasm::BBQJITImpl::BBQJIT::getGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::setGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStore):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitLoad):
* Source/JavaScriptCore/wasm/WasmBBQJIT32_64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::sizeOfType):
(JSC::Wasm::BBQJITImpl::BBQJIT::getGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::setGlobal):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCatchTableImpl):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStoreConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveConst):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitStore):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitLoad):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.h:
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCCall):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::LLIntCallee::LLIntCallee):
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::marshallLocation const):
(JSC::Wasm::WasmCallingConvention::numberOfStackResults const):
(JSC::Wasm::WasmCallingConvention::numberOfStackArguments const):
(JSC::Wasm::JSCallingConvention::marshallLocation const):
(JSC::Wasm::CCallingConventionArmThumb2::marshallLocation const):
(JSC::Wasm::CCallingConventionArmThumb2::numberOfStackResults const):
(JSC::Wasm::CCallingConventionArmThumb2::numberOfStackArguments const):
* Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp:
(JSC::Wasm::ConstExprGenerator::addConstant):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isValidHeapTypeKind):
(JSC::Wasm::heapTypeKindAsString):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.cpp:
(JSC::Wasm::HandlerInfo::initialize):
(JSC::Wasm::HandlerInfo::handlerForIndex):
* Source/JavaScriptCore/wasm/WasmHandlerInfo.h:
(JSC::Wasm::UnlinkedHandlerInfo::typeName const):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addTryTable):
(JSC::Wasm::IPIntGenerator::addThrowRef):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::ControlType::createTryTable):
(JSC::Wasm::LLIntGenerator::ControlType::isTryTable):
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addTryTable):
(JSC::Wasm::LLIntGenerator::addThrowRef):
(JSC::Wasm::LLIntGenerator::endTryTable):
(JSC::Wasm::LLIntGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::isTryTable):
(JSC::Wasm::OMGIRGenerator::ControlData::dump const):
(JSC::Wasm::OMGIRGenerator::ControlData::endTryTable):
(JSC::Wasm::OMGIRGenerator::ControlData::tryStart const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryEnd const):
(JSC::Wasm::OMGIRGenerator::ControlData::tryDepth const):
(JSC::Wasm::OMGIRGenerator::ControlData::exception const):
(JSC::Wasm::OMGIRGenerator::addTryTable):
(JSC::Wasm::OMGIRGenerator::emitCatchTableImpl):
(JSC::Wasm::OMGIRGenerator::addThrowRef):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::typeKindSizeInBytes):
* Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
(JSC::JSWebAssemblyStruct::set):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/283954@main
